viewing_party/party.py

```diff
@@ -1,23 +1,176 @@
+    
+    return user_data
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
+def get_watched_avg_rating(user_data):
+    if len(user_data["watched"]) == 0:
+        return float(0)
+
+    rating_sum = 0
+    for i in range(len(user_data["watched"])):
+        rating_sum += user_data["watched"][i]["rating"]
+    
+    avg_rating = rating_sum / len(user_data["watched"])
+    return avg_rating
 
+def get_most_watched_genre(user_data):
+    if len(user_data["watched"]) == 0:
+        return None
+
+    watched_genres = {}
+    for movie in user_data["watched"]:
+        movie_genre = movie["genre"]
+        if movie_genre in watched_genres:
+            watched_genres[movie_genre] += 1
+        else:
+            watched_genres[movie_genre] = 1
+
+    most_watched = max(watched_genres, key = watched_genres.get)
+    return most_watched
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
+def get_movie_titles(watched_list):
+    """
+    Takes in one parameter, a list of dictionaries
+    """
+    movie_titles = []
+
+    for movies in watched_list:
+        for info in movies.keys():
+            if info == "title":
+                movie_titles.append(movies["title"])
+
+    return movie_titles
+
+def make_list_of_friends_movies(user_data):
+    friends_movies = []
+    for watched_lists in user_data["friends"]:
+        titles = get_movie_titles(watched_lists["watched"])
+        friends_movies.append(titles)
+
+    return friends_movies
+
+def make_set_from_nested_list(nested_list):
+    new_set = set()
+    for inside_lists in nested_list:
+        for items in inside_lists:
+            new_set.add(items)
+    return new_set
+
+def get_unique_watched(user_data): 
+    user_movies = get_movie_titles(user_data["watched"])
+    friends_movies = make_list_of_friends_movies(user_data)
+
+    friends_movies_set = make_set_from_nested_list(friends_movies)
+    user_unique_watched = set(user_movies) - friends_movies_set
+    
+    unique_movies = []
+    for movie in user_data["watched"]:
+        if movie["title"] in user_unique_watched:
+            unique_movies.append(movie)
+    
+    return unique_movies
+
+def get_friends_unique_watched(user_data):
+    user_movies = get_movie_titles(user_data["watched"])
+    friends_movies = make_list_of_friends_movies(user_data)
+    # Combine movies together into one set
+    friends_movies_set = make_set_from_nested_list(friends_movies)
+
+    friend_unique_watched = friends_movies_set - set(user_movies)
+    
+    friend_unique_movies = []
+    for watched_lists in user_data["friends"]:
+        for movie in watched_lists["watched"]:
+            if movie["title"] in friend_unique_watched:
+                if movie in friend_unique_movies:
+                    continue
+                friend_unique_movies.append(movie)

```

Comment:
Let's look at this a little closer. Without using any of the lists created by the helper functions, we still looped through the friends list, then looped through each movie in the friend's watched.

Then we checked if the movie was in the user's watched list. If it's _not_, and the movie is _not_ already in our new `friend_unique_movies` list, then we append it

```suggestion
    for friends in user_data["friends"]:
        for movie in friends["watched"]:
            if movie["title"] not in user_data["watched"] and movie not in friend_unique_movies:
                friend_unique_movies.append(movie)
```